# FIFA World Cup Data

In this notebook, we will have a first look at the data from previous FIFA World Cup. We will also treat and validate the information present in the datasets.

## Dataset

First, let's import pandas library.

In [1]:
import pandas as pd

We then read the datasets we will analyze. These datasets were obtained from [kaggle](https://www.kaggle.com/abecklas/fifa-world-cup/) and contain information about the matches, players and cups.

In [2]:
matches = pd.read_csv('data/WorldCupMatches.csv')
players = pd.read_csv('data/WorldCupPlayers.csv')
world_cups = pd.read_csv('data/WorldCups.csv')

Let's have a look at the characteristics of each dataset imported.

### Matches dataset

As we can see below, this dataset contains general information about every World Cup match. It depicts the date, location and attendance of the game, the stage (group or knockout round), name and initials of home and away teams, half-time and final score, win conditions (decided in extra time or penalties), name of referees and round and  match ids.

In [3]:
matches.head(3)

Year              Datetime    Stage         Stadium         City  \
0  1930.0  13 Jul 1930 - 15:00   Group 1         Pocitos  Montevideo    
1  1930.0  13 Jul 1930 - 15:00   Group 4  Parque Central  Montevideo    
2  1930.0  14 Jul 1930 - 12:45   Group 2  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France              4.0              1.0         Mexico   
1            USA              3.0              0.0        Belgium   
2     Yugoslavia              2.0              1.0         Brazil   

  Win conditions  Attendance  Half-time Home Goals  Half-time Away Goals  \
0                     4444.0                   3.0                   0.0   
1                    18346.0                   2.0                   0.0   
2                    24059.0                   2.0                   0.0   

                  Referee               Assistant 1            Assistant 2  \
0  LOMBARDI Domingo (URU)     CRISTOPHE Henry (BEL)    REGO Gilberto (BRA)   
1       MACIAS Jose (ARG)  MATEUCCI Francisco (URU)  WARNKEN Alberto (CHI)   
2     TEJADA Anibal (URU)   VALLARINO Ricardo (URU)    BALWAY Thomas (FRA)   

   RoundID  MatchID Home Team Initials Away Team Initials  
0    201.0   1096.0                FRA                MEX  
1    201.0   1090.0                USA                BEL  
2    201.0   1093.0                YUG                BRA

Now, let's check the size of our dataset and if there is any empty values in it.

In [4]:
matches.shape

(4572, 20)

In [5]:
matches.info(null_counts=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4572 entries, 0 to 4571
Data columns (total 20 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Year                  852 non-null    float64
 1   Datetime              852 non-null    object 
 2   Stage                 852 non-null    object 
 3   Stadium               852 non-null    object 
 4   City                  852 non-null    object 
 5   Home Team Name        852 non-null    object 
 6   Home Team Goals       852 non-null    float64
 7   Away Team Goals       852 non-null    float64
 8   Away Team Name        852 non-null    object 
 9   Win conditions        852 non-null    object 
 10  Attendance            850 non-null    float64
 11  Half-time Home Goals  852 non-null    float64
 12  Half-time Away Goals  852 non-null    float64
 13  Referee               852 non-null    object 
 14  Assistant 1           852 non-null    object 
 15  Assistant 2          

So, it looks like we have 4572 rows, but only 852 of them are non-null. Also, there's 2 more values missing in the **Attendance** column. We'll treat those missing values later.

### Players dataset

This dataset contains data of the players that were in the line-up of all matches. We can see the player's name, team, position (goal keeper or captain), coach name and shirt number. Also, there is information about wheter the player was in line-up or substitute and any event (goal, card, substitution) related to the player that occurred in the match. Finally, matches and rounds are identified by their unique ids.

In [6]:
players.head(3)

RoundID  MatchID Team Initials           Coach Name Line-up  Shirt Number  \
0      201     1096           FRA  CAUDRON Raoul (FRA)       S             0   
1      201     1096           MEX     LUQUE Juan (MEX)       S             0   
2      201     1096           FRA  CAUDRON Raoul (FRA)       S             0   

        Player Name Position Event  
0       Alex THEPOT       GK   NaN  
1   Oscar BONFIGLIO       GK   NaN  
2  Marcel LANGILLER      NaN  G40'

It seems our dataset only has empty cells in the last two columns (**Position** and **Event**). This is ok, since only goalkeepers and captains are identified in the former and not all players were expected to be involved in events during a match.

In [7]:
players.shape

(37784, 9)

In [8]:
players.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 37784 entries, 0 to 37783
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   RoundID        37784 non-null  int64 
 1   MatchID        37784 non-null  int64 
 2   Team Initials  37784 non-null  object
 3   Coach Name     37784 non-null  object
 4   Line-up        37784 non-null  object
 5   Shirt Number   37784 non-null  int64 
 6   Player Name    37784 non-null  object
 7   Position       4143 non-null   object
 8   Event          9069 non-null   object
dtypes: int64(3), object(6)
memory usage: 2.6+ MB


### World Cups dataset

The final dataset contains general data about the world cup editions, such as the year it occurred, hosting country/countries, winner, 2nd, 3rd and 4th places, number of goals scored, participants, matches played and overall attenndance.

In [9]:
world_cups.head(3)

Year  Country   Winner      Runners-Up    Third      Fourth  GoalsScored  \
0  1930  Uruguay  Uruguay       Argentina      USA  Yugoslavia           70   
1  1934    Italy    Italy  Czechoslovakia  Germany     Austria           70   
2  1938   France    Italy         Hungary   Brazil      Sweden           84   

   QualifiedTeams  MatchesPlayed Attendance  
0              13             18    590.549  
1              16             17    363.000  
2              15             18    375.700

Great! The _world_cups_ dataset has no missing fields.

In [10]:
world_cups.shape

(20, 10)

In [11]:
world_cups.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20 entries, 0 to 19
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Year            20 non-null     int64 
 1   Country         20 non-null     object
 2   Winner          20 non-null     object
 3   Runners-Up      20 non-null     object
 4   Third           20 non-null     object
 5   Fourth          20 non-null     object
 6   GoalsScored     20 non-null     int64 
 7   QualifiedTeams  20 non-null     int64 
 8   MatchesPlayed   20 non-null     int64 
 9   Attendance      20 non-null     object
dtypes: int64(4), object(6)
memory usage: 1.7+ KB


### Data cleanse and validation

Before start analysing our data, we should treat those empty fields previously identified and look if there is any duplicated rows in our datasets. After that, just to be 100% sure of our data, we can perform some validations in our dataset in order to confirm that no relevant information is missing.

#### Treating missing data

The only dataset that we identified problems with missing values is _matches_, that contains more missing rows than actual data. Let's get rid of those rows with no data and then have a look at the empty attendance fields remaining.

In [12]:
matches = matches[matches.MatchID.notna()]

In [13]:
matches[matches.Attendance.isna()]

Year              Datetime        Stage            Stadium  \
823  2014.0  30 Jun 2014 - 17:00   Round of 16  Estadio Beira-Rio   
841  2014.0  30 Jun 2014 - 17:00   Round of 16  Estadio Beira-Rio   

              City Home Team Name  Home Team Goals  Away Team Goals  \
823  Porto Alegre         Germany              2.0              1.0   
841  Porto Alegre         Germany              2.0              1.0   

    Away Team Name                 Win conditions  Attendance  \
823        Algeria  Germany win after extra time          NaN   
841        Algeria  Germany win after extra time          NaN   

     Half-time Home Goals  Half-time Away Goals             Referee  \
823                   0.0                   0.0  RICCI Sandro (BRA)   
841                   0.0                   0.0  RICCI Sandro (BRA)   

                   Assistant 1              Assistant 2   RoundID  \
823  DE CARVALHO Emerson (BRA)  VAN GASSE Marcelo (BRA)  255951.0   
841  DE CARVALHO Emerson (BRA)  VAN GASSE Marcelo (BRA)  255951.0   

         MatchID Home Team Initials Away Team Initials  
823  300186460.0                GER                ALG  
841  300186460.0                GER                ALG

Hmmm...

Seems like the rows with missing attendance info are duplicated in our dataset. We should check if there's more duplicated rows in remove it.

In [14]:
matches[matches.duplicated()]

Year              Datetime                     Stage  \
836  2014.0  28 Jun 2014 - 13:00                Round of 16   
837  2014.0  28 Jun 2014 - 17:00                Round of 16   
838  2014.0  29 Jun 2014 - 13:00                Round of 16   
839  2014.0  29 Jun 2014 - 17:00                Round of 16   
840  2014.0  30 Jun 2014 - 13:00                Round of 16   
841  2014.0  30 Jun 2014 - 17:00                Round of 16   
842  2014.0  01 Jul 2014 - 13:00                Round of 16   
843  2014.0  01 Jul 2014 - 17:00                Round of 16   
844  2014.0  04 Jul 2014 - 13:00             Quarter-finals   
845  2014.0  04 Jul 2014 - 17:00             Quarter-finals   
846  2014.0  05 Jul 2014 - 13:00             Quarter-finals   
847  2014.0  05 Jul 2014 - 17:00             Quarter-finals   
848  2014.0  08 Jul 2014 - 17:00                Semi-finals   
849  2014.0  09 Jul 2014 - 17:00                Semi-finals   
850  2014.0  12 Jul 2014 - 17:00   Play-off for third place   
851  2014.0  13 Jul 2014 - 16:00                      Final   

                 Stadium             City Home Team Name  Home Team Goals  \
836     Estadio Mineirao  Belo Horizonte          Brazil              1.0   
837  Estadio do Maracana  Rio De Janeiro        Colombia              2.0   
838     Estadio Castelao       Fortaleza     Netherlands              2.0   
839     Arena Pernambuco          Recife      Costa Rica              1.0   
840     Estadio Nacional        Brasilia          France              2.0   
841    Estadio Beira-Rio    Porto Alegre         Germany              2.0   
842   Arena de Sao Paulo       Sao Paulo       Argentina              1.0   
843     Arena Fonte Nova        Salvador         Belgium              2.0   
844  Estadio do Maracana  Rio De Janeiro          France              0.0   
845     Estadio Castelao       Fortaleza          Brazil              2.0   
846     Estadio Nacional        Brasilia       Argentina              1.0   
847     Arena Fonte Nova        Salvador     Netherlands              0.0   
848     Estadio Mineirao  Belo Horizonte          Brazil              1.0   
849   Arena de Sao Paulo       Sao Paulo     Netherlands              0.0   
850     Estadio Nacional        Brasilia          Brazil              0.0   
851  Estadio do Maracana  Rio De Janeiro         Germany              1.0   

     Away Team Goals Away Team Name                         Win conditions  \
836              1.0          Chile       Brazil win on penalties (3 - 2)    
837              0.0        Uruguay                                          
838              1.0         Mexico                                          
839              1.0         Greece   Costa Rica win on penalties (5 - 3)    
840              0.0        Nigeria                                          
841              1.0        Algeria          Germany win after extra time    
842              0.0    Switzerland        Argentina win after extra time    
843              1.0            USA          Belgium win after extra time    
844              1.0        Germany                                          
845              1.0       Colombia                                          
846              0.0        Belgium                                          
847              0.0     Costa Rica  Netherlands win on penalties (4 - 3)    
848              7.0        Germany                                          
849              0.0      Argentina    Argentina win on penalties (2 - 4)    
850              3.0    Netherlands                                          
851              0.0      Argentina          Germany win after extra time    

     Attendance  Half-time Home Goals  Half-time Away Goals  \
836     57714.0                   0.0                   0.0   
837     73804.0                   1.0                   0.0   
838     58817.0                   0.0                   0.0   
839     41242.0                   0.0         

Seems all knockout matches from 2014 World Cup are duplicated in our dataset. Let's get rid of those bad apples.

In [15]:
matches.drop_duplicates(inplace=True)
matches.shape

(836, 20)

Done and dusted. The 16 duplicated entries are now gone.

Now, about the missing attendance field, we can infer its value by comparing the 2014 WC overall attendance with each match attendance. Let's do it!

In [16]:
world_cups.Attendance = world_cups.Attendance.apply(lambda att: eval(att.replace('.', '')))
world_cups.loc[world_cups.Year == 2014, "Attendance"]

19    3386810
Name: Attendance, dtype: int64

In [17]:
matches.groupby('Year').Attendance.sum()[2014]

3386810.0

Oops! Seems like the attendances in _world_cups_ were calculated using the _matches_ info, but we know for sure that match had a crowd. 

Searching on [Wikipedia](https://en.wikipedia.org/wiki/2014_FIFA_World_Cup?oldformat=true#Round_of_16), we can see that Germany vs. Algeria had an attendance of **43.063** fans. Let's add this info into both our datasets.

In [18]:
matches.loc[matches.Attendance.isna(), "Attendance"] = 43063
new_overall_attendance = matches.groupby('Year').Attendance.sum()[2014]
world_cups.loc[world_cups.Year == 2014, "Attendance"] = new_overall_attendance

#### Treating duplicated data

We already had a look at the duplicated data from _matches_ dataset. Now, let's check the other two.

In [19]:
world_cups[world_cups.duplicated()]

Empty DataFrame
Columns: [Year, Country, Winner, Runners-Up, Third, Fourth, GoalsScored, QualifiedTeams, MatchesPlayed, Attendance]
Index: []

In [20]:
players[players.duplicated()]

RoundID    MatchID Team Initials                 Coach Name Line-up  \
37048   255951  300186487           BRA  SCOLARI Luiz Felipe (BRA)       S   
37049   255951  300186487           CHI       SAMPAOLI Jorge (ARG)       S   
37050   255951  300186487           BRA  SCOLARI Luiz Felipe (BRA)       S   
37051   255951  300186487           CHI       SAMPAOLI Jorge (ARG)       S   
37052   255951  300186487           BRA  SCOLARI Luiz Felipe (BRA)       S   
...        ...        ...           ...                        ...     ...   
37779   255959  300186501           ARG    SABELLA Alejandro (ARG)       N   
37780   255959  300186501           GER         LOEW Joachim (GER)       N   
37781   255959  300186501           ARG    SABELLA Alejandro (ARG)       N   
37782   255959  300186501           GER         LOEW Joachim (GER)       N   
37783   255959  300186501           ARG    SABELLA Alejandro (ARG)       N   

       Shirt Number  Player Name Position       Event  
37048            12  JULIO CESAR       GK         NaN  
37049             1     C. BRAVO      GKC         NaN  
37050             2   DANI ALVES      NaN       Y105'  
37051             2         MENA      NaN        Y17'  
37052             3      T SILVA        C         NaN  
...             ...          ...      ...         ...  
37779            19      ALVAREZ      NaN         NaN  
37780             6      KHEDIRA      NaN         NaN  
37781            20       AGUERO      NaN  IH46' Y65'  
37782            21      MUSTAFI      NaN         NaN  
37783            23      BASANTA      NaN         NaN  

[736 rows x 9 columns]

So, we also have a few hundred duplicated rows in the _players_ dataset, but none in _world_cups_. Let's just drop those rows and move on to our validations.

In [21]:
players.drop_duplicates(inplace=True)

#### Data validation

In this stage, we wanna check if the information on the three datasets match and, if not, try and correct them. 

First, let's create a function to compare information from two different sources. This function will be later used to validate data accross the different datasets we have.

In [22]:
def validate_data(series1, series2, data_label='info'):
    print(f'--- {data_label} ---')
    
    if len(series1.index) != len(series2.index):
        print('* Series must have the same lenght')
        return None
    
    if not all(series1.index == series2.index):
        print('* Series must have the same indices')
        return None
    
    aux = True
    for index in series1.index:
        val1, val2 = series1.loc[index], series2.loc[index]
        if val1 != val2:
            print(f'* Mismatch in value of \'{data_label}\' for index {index}: {val1} != {val2}')
            aux = False
    
    if aux:
        print('* OK!')

##### World Cups vs. Matches

Let's compare the overall information from _world_cups_ with the match data from _matches_.

In [23]:
# Get number of matches, goals and participants for each world cup edition from matches dataset
edition_matches = matches.groupby('Year')
n_match_matches = edition_matches.MatchID.nunique()
n_goals_matches = edition_matches[['Home Team Goals', 'Away Team Goals']].sum().sum(axis=1)

edition_teams_matches = edition_matches[['Home Team Name', 'Away Team Name']]
n_teams_matches = edition_teams_matches.apply(lambda x: len(set(x['Home Team Name']).union(set(x['Away Team Name']))))

# Get number of matches, goals and participants for each world cup edition from world_cups dataset
wcups_ = world_cups.set_index('Year')
n_match_wcups = wcups_.MatchesPlayed
n_goals_wcups = wcups_.GoalsScored
n_teams_wcups = wcups_.QualifiedTeams

In [24]:
validate_data(n_match_matches, n_match_wcups, data_label='# of matches')
validate_data(n_goals_matches, n_goals_wcups, data_label='# of goals')
validate_data(n_teams_matches, n_teams_wcups, data_label='# of teams')

--- # of matches ---
* OK!
--- # of goals ---
* OK!
--- # of teams ---
* Mismatch in value of '# of teams' for index 1978.0: 17 != 16
* Mismatch in value of '# of teams' for index 2006.0: 33 != 32


We have a mismatch on the number of team in two editions: 1978 and 2006. By looking at the teams of those two editions, we can see that Iran appears twice (as 'Iran' and 'IR Iran').

In [25]:
edition_teams_matches = matches.groupby('Year')[['Home Team Name', 'Away Team Name']]
print('1978:', edition_teams_matches.apply(lambda x: set(x['Home Team Name']).union(set(x['Away Team Name'])))[1978])
print('2006:', edition_teams_matches.apply(lambda x: set(x['Home Team Name']).union(set(x['Away Team Name'])))[2006])

1978: {'Spain', 'Italy', 'IR Iran', 'Sweden', 'Hungary', 'Scotland', 'Tunisia', 'Germany FR', 'Peru', 'Netherlands', 'Argentina', 'Mexico', 'Austria', 'Poland', 'France', 'Brazil', 'Iran'}
2006: {'Paraguay', 'rn">Serbia and Montenegro', 'Korea Republic', "C�te d'Ivoire", 'Sweden', 'USA', 'Tunisia', 'Czech Republic', 'Switzerland', 'Netherlands', 'Australia', 'rn">Trinidad and Tobago', 'Ukraine', 'Costa Rica', 'France', 'Ecuador', 'Portugal', 'Spain', 'Italy', 'Ghana', 'Germany', 'IR Iran', 'Japan', 'Iran', 'Argentina', 'Croatia', 'Angola', 'England', 'Saudi Arabia', 'Poland', 'Togo', 'Brazil', 'Mexico'}


Let's replace 'IR Iran' with 'Iran' in order to solve this mismatch.

In [26]:
matches.replace('IR Iran', 'Iran', inplace=True)
edition_teams_matches = edition_matches[['Home Team Name', 'Away Team Name']]
n_teams_matches_fixed = edition_teams_matches.apply(lambda x: len(set(x['Home Team Name']).union(set(x['Away Team Name']))))
validate_data(n_teams_matches_fixed, n_teams_wcups, data_label='# of teams')

--- # of teams ---
* OK!


##### Players

Let's confirm that the teams that played each match identified by MatchID are the same on both _players_ and _matches_ datasets.

In [27]:
# Get initial of teams for world cup match from players dataset
teams_match_players = players.groupby('MatchID')['Team Initials'].apply(set)

# Get initial of teams for world cup match from matches dataset
teams_match_matches = matches.groupby('MatchID').apply(lambda x: set(x['Home Team Initials']).union(set(x['Away Team Initials'])))

validate_data(teams_match_players, teams_match_matches, data_label='Teams on matches')

--- Teams on matches ---
* OK!


Hooray!

Finally, as we can see below, for all matches, every team had at least 11 players listed.

In [28]:
all(players.groupby(['MatchID', 'Team Initials']).nunique()['Player Name'] >= 11)

True

### Export datasets

After cleaning and validating our data, let's export them as csv files that will be later used for exploration and model training

In [29]:
matches.to_csv('data/WorldCupMatches_clean.csv', index=False)
players.to_csv('data/WorldCupPlayers_clean.csv', index=False)
world_cups.to_csv('data/WorldCups_clean.csv', index=False)